src/datahike/writer.cljc

```diff
@@ -40,77 +40,90 @@
-                                               {:type       :writer-error-during-invocation
-                                                :invocation invocation
-                                                :connection connection
-                                                :error      e})
-                                      e))
-                              :error))]
-                (when-not (= res :error)
-                  (when (> (count commit-queue-buffer) (/ commit-queue-size 2))
-                    (log/warn "Commit queue buffer more than 50% full, "
-                              (count commit-queue-buffer) "of" commit-queue-size  " filled."
-                              "Throttling transaction processing. Reduce transaction frequency and check your storage throughput.")
-                    (<! (timeout 50)))
-                  (put! commit-queue [res callback])))
-              (recur))
-            (do
-              (close! commit-queue)
-              (log/debug "Writer thread gracefully closed"))))
+                                      (put! callback
+                                            (if (= (type e) NullPointerException)
+                                              (ex-info "Null pointer encountered in invocation. Connection may have been invalidated, e.g. through db deletion, and needs to be released everywhere."
+                                                       {:type       :writer-error-during-invocation
+                                                        :invocation invocation
+                                                        :connection connection
+                                                        :error      e})
+                                              e))
+                                      :error))]
+                        (if-not (= res :error)
+                          (do
+                            (when (> (count commit-queue-buffer) (/ commit-queue-size 2))
+                              (log/warn "Commit queue buffer more than 50% full, "
+                                        (count commit-queue-buffer) "of" commit-queue-size  " filled."
+                                        "Throttling transaction processing. Reduce transaction frequency and check your storage throughput.")
+                              (<! (timeout 50)))
+                            (put! commit-queue [res callback])
+                            (recur (:db-after res)))
+                          (recur old))))
+                    (do
+                      (close! commit-queue)
+                      (log/debug "Writer thread gracefully closed")))))
         ;; commit loop
-        (go-loop [tx (<?- commit-queue)]
-          (when tx
-            (let [txs (atom [tx])]
+        (go-try S
+                (loop [tx (<?- commit-queue)]
+                  (when tx
+                    (let [txs (atom [tx])]
```

Comment:
Yes, beautiful. Unfortunately I cannot always use transducers since sometimes I need to async operations in the transducer function itself, but here it works. Thanks!

Reply:
Nice that it worked! Yes, it was a while since I tried using core.async so I forgot what is possible and what is not.